### ISSUE TITLE
How can I make requests in response to updates?

### ISSUE THREAD

**Issue by zumoshi** [nickoala/telepot#63]
Hello,
as per telegram faq , you can [make request in response to updates](https://core.telegram.org/bots/faq#how-can-i-make-requests-in-response-to-updates), removing the need for another request. while this may not be optimal for sending messages or other methods due to inability to know that such a request was successful or get its result, but it can help speed up inline queries.
the way i can see it we can implant it in one of these ways:

- create a new function for each method that instead of making the request returns the json to be sent as response.
- add a parameter to Bot class `__init__` to chose its behavior between making a post request or returning json
- add a parameter to each function method to define behavior , something like `return_request=False`

i prefer the last one and i can make a pull request for it if you think it is a good addition to telepot,
i'm open to other ideas on how to add this feature though.

**Comment by nickoala** [comment]
It seems you missed the fact that it only applies to webhook ... which renders your suggestions ... moot  :smile:

**Comment by zumoshi** [comment]
i am aware of that, in my use case that option would be used in addition to flask/cherrypy/cgi/....
when if `return_request` is true instead of returning the response from bot api it would just return a string containing the json to be sent as response. than depending on how you set up the webhook you either return it or print it or something.

take this flask skeleton for example:
```python
from flask import Flask, request

app = Flask(__name__)

@app.route('/webhook_path', methods=['GET', 'POST'])
def pass_update():
    update_queue.put(request.data)  # dump data to queue
    return 'OK'
```

here instead of returning OK we will return the json : 
```python
from flask import Flask, request

app = Flask(__name__)

@app.route('/webhook_path', methods=['GET', 'POST'])
def pass_update():
    query_id, from_id, query = glance(request.data['inline_query'], flavor='inline')
    return bot.answerInlineQuery(query_id, [InlineQueryResultArticle(
                id='1',
                text=query ,
                title='Current Unix timestamp',
    )], return_response=True)
```

of course in a real application we have to check if it is an inline query before doing this and make sure its thread safe and all that, but for the sake of the example i just assumed flask is running a single thread and bot only receives inline queries.

it is not the perfect solution because of many reasons, not being able to use queues and not getting the result and all, but we are not forcing it, just as an extra way of doing it and to support all telegram bot api features.

**Comment by nickoala** [comment]
Clever :+1: 

After a few days considering different options, I decide not to add this functionality. The benefit is just not clear. Quicker response? Yes, in theory. How much quicker in practice? Nobody knows. It just seems too much ado for not much.

Instead of overriding `_post`, I suggest you look at `_rectify`. This function takes a dict (presumably holding all the params), eliminates all `None` entries, flattens all namedtuples/dicts/lists to JSON strings, and produces a simplified dict whose values can be passed directly to Bot API. Using that, you should be able to construct an HTTP response quite easily. And because of this workaround, there is even less incentive for me to incorporate this second behavior.

**Comment by zumoshi** [comment]
as for `_rectify` i saw it while looking for a work around, however patching that won't prevent a request to be made to telegram bot api. i needed to prevent the request and return the json instead.

but i would still prefer to have this , i understand if it may not be appealing enough to modify main Bot class, but what about a wrapper class, like `Sender` class? that would return json request for methods instead of calling them?

some other reasons why this may be needed:

- if outbound connections are blocked by the firewall (this method still 
- if connection pooling is not possible (e.g. in cgi when app is run once per request than terminated) without keep-alive overhead of ssl handshake is considerable, but this prevents it

also this workaround is not complete, it can't send files (while its possible to do so with the right headers), also has some other problems with json (sometimes json created by it includes a nested array as json string instead of array), so its not easy to get this right. 

i think this deserves at the very least an example of patching `_post` the right way in docs, if you still don't want to embed it in the telepot.

**Comment by nickoala** [comment]
By mentioning `_rectify()`, I meant you could construct the response using `_rectify()`, then jsonize it:
```python
@app.route('/webhook_path', methods=['GET', 'POST'])
def pass_update():
    query_id, from_id, query = glance(request.data['inline_query'], flavor='inline')

    p = {'method': 'answerInlineQuery',
         'query_id': query_id,
         'results': [InlineQueryResult(...), ...]}

    # this would flatten the `results` list into a JSON string
    p = telepot._rectify(p)

    return json.dumps(p)
```

I mean, constructing the response this way is not much harder than calling `bot.answerInlineQuery(...)`.

May I also ask: how can I send files with "the right headers" as you said? Isn't the response restricted to be a JSON string?

I can't think of a way to patch `_post()` while keeping other methods working normally (e.g. download a file while telling the user you are downloading it, as you suggested). `_post()` is too interwoven into the bot's working. You can say it's a lack of foresight on my part.

Making a wrapper is always doable. The problem is whether I want to do it, or whether I want to include it in telepot. In fact, with all those helper functions in telepot, it's not hard to write a brand new class (e.g. a subclass of `_BotBase`) whose sole purpose is to transform method calls into JSON equivalents:
```python
def JsonBot(_BotBase):
    def sendMessage(self, ...):
        p = _strip(locals())
        p['method'] = 'sendMessage'
        return json.dumps(_rectify(p))

    def send*** ...
```

After staring at these few lines of code for a moment, I realize all methods would be identical (except for the method name), which goes back to my original suggestion: why don't you just call `json.dumps(_rectify({...}))`?

I deliberately ignore those cases where more manipulations of parameters are needed, for example when sending files, but as I said above, I don't see how you could send a file in response to a webhook request anyway.

Did I missing anything?

**Comment by zumoshi** [comment]
in my experience flattening the `results` actually causes an error instead of solving one. when sending as post you were sending it as form data, so each key must have had a string as value , but when the whole thing is a json string , embedding a json as string inside the root json causes an error. (which we don't get since you can't retrieve error messages but the bot didn't work until i patched out that part).

as for uploading (or downloading? since api is technically downloading it) i didn't test it myself but from api help:

> If you're using webhooks, you can perform a request to the Bot API while sending an answer to the webhook. Use either application/json or application/x-www-form-urlencoded or **multipart/form-data** response content type for passing parameters. 

source: https://core.telegram.org/bots/api#making-requests-when-getting-updates

i assumed it means we can send files this way, but have to test it to be sure.

--

for your suggestion , creating the request object manually kind of defeats the purpose of using telepot altogether, i could make an object and json dump it without needing telepot. but i like helper functions, specially if they could predict errors (i may open a separate issue about this soon) which would help a lot since we won't get errors from telegram.

**Comment by nickoala** [comment]
Your link above points to this link:  

https://core.telegram.org/bots/faq#how-can-i-make-requests-in-response-to-updates

which, if i understand correctly, requires a POST request to send a file, just as it always does.

You said "creating the request object manually kind of defeats the purpose of using telepot". I agree. That means, if you want to obtain a JSON string, you may as well not use telepot, because it's almost as simple as using telepot.

Nonetheless, the helper functions would be helpful. As for the error caused by `_rectify`, I have not experienced that myself. Let me try using it in a webhook response one day to verify what you said.

**Comment by nickoala** [comment]
Ok, done it. First, a webhook that responds to **chat messages** only. Note that you can't use the raw request data right away, have to parse it yourself.

```python
import sys
import json
from flask import Flask, request, Response
import telepot

TOKEN = sys.argv[1]
PORT = int(sys.argv[2])
URL = sys.argv[3]

app = Flask(__name__)
bot = telepot.Bot(TOKEN)

@app.route('/abc', methods=['GET', 'POST'])
def pass_update():
    msg = json.loads(request.data)['message']

    r = {'method': 'sendMessage',
         'chat_id': msg['chat']['id'],
         'text': 'Quick response'}

    return Response(json.dumps(telepot._rectify(r)), mimetype='application/json')


if __name__ == '__main__':
    bot.setWebhook(URL)
    app.run(port=PORT, debug=True)
```

Then, a webhook that responds to **inline query**.

```python
import sys
import json
from flask import Flask, request, Response
import telepot

TOKEN = sys.argv[1]
PORT = int(sys.argv[2])
URL = sys.argv[3]

app = Flask(__name__)
bot = telepot.Bot(TOKEN)

@app.route('/abc', methods=['GET', 'POST'])
def pass_update():
    iq = json.loads(request.data)['inline_query']
    query_id = iq['id']

    articles = [{'type': 'article',
                    'id': 'abc', 'title': 'ABC', 'message_text': 'Good morning'}]

    r = {'method': 'answerInlineQuery',
         'inline_query_id': query_id,
         'results': articles}

    return Response(json.dumps(telepot._rectify(r)), mimetype='application/json')


if __name__ == '__main__':
    bot.setWebhook(URL)
    app.run(port=PORT, debug=True)
```

Assuming that you can't upload files with a direct response, anything you can do in that response has to be quite simple and straight-forward. With the help of `_rectify`, it is so simple and straight-forward that you may as well construct the response yourself, without going through telepot.

**Comment by zumoshi** [comment]
i just tested it and you can in fact send files using this method. telegram has not exactly implanted the http protocol the way specification dictates (as you have encountered yourself in the filename issue)

it just simply treats what was send as response as if it was send in a request. i've send the file using the following code:
```python
#!/home/bor691/pyboz/bin/python
from requests.models import Request
import telepot, json, sys


def handle(update):
    bot = telepot.Bot('TOKEN_HERE')
    if 'message' in update:
        msg = update['message']
        content_type, chat_type, chat_id = telepot.glance(msg)
        if msg['text'] == '/send_file':
            bot.sendPhoto(chat_id, open('./test.png', 'rb'))
        else:
            bot.sendMessage(chat_id, 'Hi')


def patch():
    def _print(url, **kwargs):
        # here i add method to data
        kwargs['data'].update({'method': url.split('/')[-1]})
        # here i fake a request, to leverage requests library and
        # not having to manually create the response body
        r = Request(
            'post',             # method must be post
            'http://anything',  # this doesn't matter as long as it starts with http://
                                # as to not get an exception, we are only interested
                                # in the generated body
            # this is to remove timeout:
            **dict((k, v) for k, v in kwargs.items() if k in ['files', 'data'])
        )
        # create body and headers using requests
        p = r.prepare()

        for header, value in p.headers.items():
            # print headers , since print() can't send bytes i used sys.stdout.buffer
            sys.stdout.buffer.write(':'.join([header, ' ' + value]).encode())
            sys.stdout.buffer.write('\r\n'.encode())
        # the extra linefeed that marks the end of headers
        sys.stdout.buffer.write('\r\n'.encode())
        # requests sends str in case of url-encoded and bytes for multipart
        if isinstance(p.body, str):
            sys.stdout.buffer.write(p.body.encode())
        else:
            sys.stdout.buffer.write(p.body)

    telepot._post = _print


patch()
handle(json.load(sys.stdin))

```

although your examples are helpful but i prefer to use telepot methods. if you don't want to implant it i will create a wrapper class and use it myself. if you were interested i can make a pull request with that wrapper class when it got ready. 

but even if you don't want to put it in the library at least update the Bot class to have `_post` function as part of class itself, so i can override it in a subclass, that'll make the wrapper as small as a subclass with the method i provided above, or different subclasses, one that would dump the request in stdout while another returns a tuple of `(headers, bytes)` for more general use case outside of cgi. `_post` is used only inside Bot class anyways why does it have to be outside it ?

**Comment by nickoala** [comment]
Ok. I am quite impressed that you found a way to send files this way. Haven't tested it myself, but will take your words for it.

One thing I want to make sure: by responding this way, you can only *do one thing*, right? For example, you cannot send *two* messages, nor can you send a message and then send a file. Is that correct?

**Comment by zumoshi** [comment]
yes, only a single request. but you can for example send a file with caption instead of a text message and a file.

have you considered my suggestion to move `_post` function inside `Bot` class? that way instead of monkey patching i can subclass bot class and override `_post` method to easily achieve ability to send requests both ways. (using instances of normal bot and sub classed bot)

**Comment by nickoala** [comment]
By the way, why don't you create a Python package that is designed specifically for generating webhook responses? You may think the `Bot` class (thus telepot) is a good basis to start from, but it is not. Making a POST request is fundamentally different from making a webhook response. Your idea of overriding `_post` is just a quick trick. not a proper way to doing things in my mind.

**Comment by zumoshi** [comment]
i understand you hesitation, because once this is in the framework changing it would be hard due having to keep backward compatibility. 

however i disagree with calling overriding `_post` in `Bot` class a dirty way. it is understandable that what `_post` does is unrelated to the nature of bot, but you have admitted yourself that it was not elegant to use `_post` all over the bot and having it interwoven. this makes it hard to change it for extensibility  purposes. (even if i just want to use a different module than requests to actually do post, or queue the requests in a global queue for preventing 429 errors, etc...)

you are saying that webhook is a separate concept from the telegram bot, i would argue a Rest Api client is also as separate as webhook. what `Bot` class should do is to create an object for telegram bot api, any means of getting that `object` to telegram api is a separate concept whether it be a Post request or printing it for cgi or sending it as a response.

so, what if instead of moving `_post` inside Bot class, we add another method (e.g. `_request`) to Bot class and call that from all methods. we can call `_post` over there for the default Bot class to maintain compatibility. this way one can override the behavior and use an alternative way of sending the requests to telegram, yet we have not brought unrelated methods that work independently to Bot class.

we can than have different implantation of subclasses of Bot class in separate modules that depend on telepot, if anyone was interested he/she can maintaining them separately, i only need one for cgi at the moment.

**Comment by nickoala** [comment]
*As of this moment*, I am of the opinion that, if you are willing to make sure the code work on, and keep them working on:
- Flask
- Django
- CherryPy
- old-fashioned CGI
- what else?

I might be willing to include that in telepot. I am simply not ready to take on that responsibility.

**Comment by nickoala** [comment]
> instead of moving _post inside Bot class, we add another method (e.g. _request) to Bot class and call that from all methods ...

Ahhhhh, this might the best compromise! Yes yes yes ...... again, let me sleep on it for a few nights. I don't like making rush decisions.

**Comment by nickoala** [comment]
In 8.0, all API requests go through `bot._api_request()`. You can override that one method in the subclass.
